### PR TITLE
[sqllab] Help sqllab forget query history

### PR DESCRIPTION
**Pull request by timifasubaa**
In the sqllab asynchronous polling logic, it looks through the query history and continues to poll if there is a query in the history that is not successful. 

This PR makes it look only at the last 6 hours of queries when checking for their states. The rationale is that a query that was started 6 hours ago is probably dead or problematic and should be rerun. 

@graceguo-supercat @mistercrunch @john-bodley

### CODE REVIEW

**Comment by mistercrunch** [comment]
I've been meaning to do something like this. I wonder if the state should be altered though, like setting the state to `failed` or `timeout`. Also build is failing...

**Comment by john-bodley** [comment]
I agree with @mistercrunch it would be great to change the state in the `queries` table for clarity and auditing purposes. Note this would actually be a backend change as opposed to a frontend change.

**Comment by timifasubaa** [comment]
Okay, I will change the logic to set all pengind or running queries more than 6 hours old to failed.

**Comment by john-bodley** [comment]
@timifasubaa I wonder if `timedout` is more descriptive that `failed` since the later can mean multiple things.

**Comment by timifasubaa** [comment]
There is a lot of repeated logic with stop_query for the new timedout query logic.

File: superset/assets/src/SqlLab/components/QueryAutoRefresh.jsx [hunk 1]
```diff
@@ -16,12 +17,23 @@ class QueryAutoRefresh extends React.PureComponent {
   componentWillUnmount() {
     this.stopTimer();
   }
+  timeoutOldQueries() {
```
Comment (graceguo-supercat) [inline superset/assets/src/SqlLab/components/QueryAutoRefresh.jsx]: This logic should not be in front-end (browser-side). 
Query table knows each query started time. it doesn't need front-end sends ajax call to set query state. 
when front-end ask a query's status, server side should first check the query's start time, if

- longer than 6 hours old: set it timeout and return timeout state.
- < 6 hours: return its current status.

**Comment by timifasubaa** [comment]
I changed the logic to update the db's sate to timeout before it responds to polls.

File: superset/views/core.py [hunk 2]
```diff
@@ -2532,6 +2532,20 @@ def queries(self, last_updated_ms):
         # UTC date time, same that is stored in the DB.
         last_updated_dt = utils.EPOCH + timedelta(seconds=last_updated_ms_int / 1000)
 
+        # Before you run the select query, also update the query.
+
+        now = int(round(time.time() * 1000))
+
+        update(Query).where(Query.user_id == g.user.get_id()).\
```
Comment (mistercrunch) [inline superset/views/core.py]: Nope, we can't run an update every time we poll, this runs hundreds of times per second at Airbnb. This won't work.

**Review by mistercrunch — CHANGES_REQUESTED** [review]
Can't run updates while polling.

File: superset/views/core.py (hunk 2 above)
Comment (timifasubaa) [reply superset/views/core.py]: @mistercrunch  Made updated it to only run the query if it definitely has some query in a bad state that's been there for over the timeout period.

**Comment by timifasubaa** [comment]
@john-bodley @mistercrunch @graceguo-supercat ready for review

**Comment by timifasubaa** [comment]
PING

Quick update on the state of the logic. 
What is really happening is that the frontend ceases to poll for queries older than 6 hours old. 
This is because on polls, it checks for all queries by that user whose state has changed after the last updated time. Meaning if a query is on pending for a very long time (> 6hours) and another query that's running at the same time changes state. We will no longer poll and also no longer receive any more updates. 

Is this backend code worth having? Especially since it may trick people into believing we actually timeout the queries when the query will most likely not be updated to timeout.

**Comment by graceguo-supercat** [comment]
current auto query refresh component work like this:

- in client-side, from all queries in local storage, it find out 
  - status in `'running', 'started', 'pending', 'fetching'`, and
  - started in < 6 hours
- if has any queries, polling with latest updated timestamp
- if there is no such queries, stop polling.

- in server-side, for every poll request, we get all queries
  - by user_id, and
  - changed after updated timestamp
- if some query went wrong, and query status kept as `running` but never got to updated after 6 hours, when client-side polling with a newer timestamp (< 6 hours), this query will not have a chance to be set as `timeout` state. But it won't cause client-side keep polling because now client-side added timeout threshold.

this PR LGTM. but will not merge until @mistercrunch feel comfortable. Thank you!

**Comment by timifasubaa** [comment]
PING @mistercrunch @john-bodley

File: superset/assets/src/SqlLab/components/QueryAutoRefresh.jsx [hunk 3]
```diff
@@ -19,10 +19,12 @@ class QueryAutoRefresh extends React.PureComponent {
...
     const { queries } = this.props;
-    const queryKeys = Object.keys(queries);
-    const queriesAsArray = queryKeys.map(key => queries[key]);
-    return queriesAsArray.some(q =>
-      ['running', 'started', 'pending', 'fetching'].indexOf(q.state) >= 0);
+    const now = new Date().getTime();
+    return Object.values(queries)
+      .some(
+        q => ['running', 'started', 'pending', 'fetching'].indexOf(q.state) >= 0 &&
+        now - q.startDttm < 21600000,
```
Comment (mistercrunch) [inline superset/assets/src/SqlLab/components/QueryAutoRefresh.jsx]: No hard coded values please. Put it in `src/SqlLab/constants.js`

**Review by mistercrunch — CHANGES_REQUESTED** [review]

File: superset/assets/src/SqlLab/components/QueryAutoRefresh.jsx (hunk 3 above)
Comment (timifasubaa) [reply superset/assets/src/SqlLab/components/QueryAutoRefresh.jsx]: Done

**Review by mistercrunch — APPROVED** [review]

**Pull request merged**